fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/QueryPlan.java

```diff
@@ -0,0 +1,147 @@
+ * {@link com.apple.foundationdb.record.query.plan.QueryPlanner QueryPlanner}s always return {@code RecordQueryPlan}s,
+ * but the query plan tree might contain intermediate plans that return something other than records.
+ *
+ * @param <T> the type of element produced by executing this plan
+ */
+@API(API.Status.EXPERIMENTAL)
+public interface QueryPlan<T> extends PlanHashable, RelationalPlannerExpression {
+
+    /**
+     * Execute this query plan.
+     * @param store record store from which to fetch records
+     * @param context evaluation context containing parameter bindings
+     * @param continuation continuation from a previous execution of this same plan
+     * @param executeProperties limits on execution
+     * @return a cursor of records that match the query criteria
+     */
+    @Nonnull
+    RecordCursor<T> execute(@Nonnull FDBRecordStore store, @Nonnull EvaluationContext context,
+                            @Nullable byte[] continuation, @Nonnull ExecuteProperties executeProperties);
+
+    /**
+     * Execute this query plan.
+     * @param store record store from which to fetch records
+     * @return a cursor of records that match the query criteria
+     */
+    @Nonnull
+    default RecordCursor<T> execute(@Nonnull FDBRecordStore store) {
+        return execute(store, EvaluationContext.EMPTY);
+    }
+
+    /**
+     * Execute this query plan.
+     * @param store record store to access
+     * @param context evaluation context containing parameter bindings
+     * @return a cursor of records that match the query criteria
+     */
+    @Nonnull
+    default RecordCursor<T> execute(@Nonnull FDBRecordStore store, @Nonnull EvaluationContext context) {
+        return execute(store, context, null, ExecuteProperties.SERIAL_EXECUTE);
+    }
+
+    /**
+     * Indicates whether this plan will return values in "reverse" order from the
+     * natural order of results of this plan. The return value is <code>true</code> if the plan
+     * returns elements in descending order and <code>false</code> if the elements are
+     * returned in ascending order.
+     * @return <code>true</code> if this plan returns elements in reverse order.
+     */
+    boolean isReverse();
+
+    /**
+     * Indicates whether this plan (or one of its components) scans at least a subset of the record range directly
+     * rather than going through a secondary index.
+     * A plan may only scan over a subset of all records. Someone might not use any secondary indexes explicitly, but
+     * they might make use of the primary key index. For example, if they had a compound primary key, and they issued
+     * a query for all records that had some value for the first element of their primary key, the planner will produce
+     * a plan which scans over a subset of all records.
+     * @return <code>true</code> if this plan (or one of its components) scans at least a subset of the records directly.
+     */
+    boolean hasRecordScan();
```

Comment:
Should `hasRecordScan()` and `hasFullRecordScan()` go on `RecordQueryPlan` rather than `QueryPlan<T>`? I'm not sure myself but it's a bit strange to have such an explicit reference to a "record" in `QueryPlan<T>`.

Reply:
I think those two methods here are fine. The idea is that you might one day have, say, an intersection plan with generic `QueryPlan` children, and the only way for the `IntersectionPlan` to know if it has a records scan is to ask its children. I suppose a, like, `PlanVisitor` scheme could do the same thing, but as long as we're supporting `hasRecordScan`, I think adding it to `QueryPlan` is fine.